lib/wibox/widget/checkbox.lua

```diff
@@ -0,0 +1,253 @@
+    if border_color or main_color then
+        cr:set_source(color(border_color or main_color))
+    end
+
+    local wa = outline_workarea(self, size, size)
+    cr:translate(wa.x, wa.y)
+    background_shape(cr, wa.width, wa.height)
+    cr:set_line_width(border_width)
+
+    if bg then
+        cr:save()
+        cr:set_source(color(bg))
+        cr:fill_preserve()
+        cr:restore()
+    end
+
+    cr:stroke()
+
+    cr:translate(-wa.x, -wa.y)
+
+    -- Draw the checked part
+    if self._private.checked then
+        local col = self:get_check_color() or main_color
+        border_color = self:get_check_border_color()
+        border_width = self:get_check_border_width() or 0
+        local check_shape = self:get_check_shape() or background_shape
+
+        wa = content_workarea(self, size, size)
+        cr:translate(wa.x, wa.y)
+
+        check_shape(cr, wa.width, wa.height)
+
+        if col then
+            cr:set_source(color(col))
+        end
+
+        if border_width > 0 then
+            cr:fill_preserve()
+            cr:set_line_width(border_width)
+            cr:set_source(color(border_color))
+            cr:stroke()
+        else
+            cr:fill()
+        end
+    end
+end
+
+local function fit(box, context, w, h)
+    local size = math.min(w, h)
+    return size, size
+end
+
+--- If the checkbox is checked.
+-- @property checked
+-- @param boolean
+
+for _, prop in ipairs {"border_width", "bg", "border_color", "check_border_color",
+    "check_border_width", "check_color", "shape", "check_shape", "paddings",
+    "checked", "color" } do
+    checkbox["set_"..prop] = function(self, value)
+        self._private[prop] = value
+        self:emit_signal("property::"..prop)
+        self:emit_signal("widget::redraw_needed")
+    end
+    checkbox["get_"..prop] = function(self)
+        return self._private[prop] or beautiful["checkbox_"..prop]
+    end
+end
+
+--- The checkbox color.
+-- @property color
+
+function checkbox:set_paddings(val)
+    self._private.paddings = type(val) == "number" and {
+        left   = val,
+        right  = val,
+        top    = val,
+        bottom = val,
+    } or val or {}
+    self:emit_signal("property::paddings")
+    self:emit_signal("widget::redraw_needed")
+end
+
+local function new(checked, args)
+    checked, args = checked or false, args or {}
+
+    local ret = base.make_widget(nil, nil, {
+        enable_properties = true,
+    })
+
+    util.table.crush(ret, checkbox)
+
+    ret._private.style = args.style or beautiful.checkbox_style or "default"
+    ret._private.checked = checked
+    ret._private.color = args.color and color(args.color) or nil
+
+    rawset(ret, "fit" , fit )
+    rawset(ret, "draw", draw)
+
+    return ret
+end
+
+--@DOC_widget_COMMON@
+
+--@DOC_object_COMMON@
+
+return setmetatable({}, { __call = function(_, ...) return new(...) end})
+
+-- kate: space-indent on; indent-width 4; replace-tabs on;
```

Comment:
Removing vim modelines would also be fine with me if many people are against them. We only have them because JD added them "Back Then" (tm).